userguide/tutorials/payment_plugin.adoc

```diff
@@ -1,61 +1,575 @@
+
+|===
+|Status | Status Description
+
+|PROCESSED
+|Indicates that the payment is successful
+|ERROR
+|Indicates that the payment is rejected by the gateway (insufficient funds, fails AVS check, fraud detected, etc.)
+|PENDING
+|Indicates that the payment requires a completion step (3D-S verification, HPP, etc.)
+|CANCELED
+|Indicates that the gateway wasn't contacted (DNS error, SSL handshake error, socket connect timeout, etc.)
+|UNDEFINED
+|Should be used for all other cases (socket read timeout, 500 returned, etc.)
+|===
+
+
+Kill Bill has a https://docs.killbill.io/latest/userguide_payment.html#_janitor[Janitor] system in place that attempts to fix *PENDING* and *UNKNOWN* states . It polls the plugin via `PaymentPluginApi#getPaymentInfo` method. If the plugin subsequently returns *PROCESSED*, the Janitor updates the internal payment state as well as invoice balance, etc.) accordingly.
+
+The Janitor matches the internal transactions against plugin transactions via the *transaction id*, so make sure `PaymentTransactionInfoPlugin#getKbTransactionPaymentId` is correctly implemented.
+
+You should try to avoid *UNDEFINED* as much as possible, because it is the only case where Kill Bill cannot retry payments (since the payment may or may not have happened).
+
+
+== GoCardless Plugin Tutorial
+
+In order to demonstrate creating a payment plugin, we will be creating a Kill Bill payment plugin for https://gocardless.com/[GoCardless]. GoCardless allows direct debit from customer's bank accounts. It requires a customer to set up a mandate the first time, after which it directly collects payments against the mandate. 
+
+GoCardless provides a https://developer.gocardless.com/getting-started/api/making-your-first-request/#setting-up-your-client-library[client library]. We will be using this library to integrate GoCardless with Kill Bill. For the sake of simplicity, we will be creating a very basic plugin that can only process payments. Refunds, credits and other plugin functionality will currently not be implemented.
+
+The complete code for this tutorial is available at https://github.com/killbill/killbill-gocardless-example-plugin[Github]
+
+=== How GoCardless works
+
+The first step in GoCardless would be adding a customer and setting up a payment mandate. GoCardless supports three ways to do this as explained https://developer.gocardless.com/getting-started/api/adding-your-first-customer/[here]. In this tutorial, we will be using the *hosted payment page* approach.  This includes the following steps as explained https://developer.gocardless.com/getting-started/api/adding-your-first-customer/#setting-up-direct-debit-mandates-using-hosted-payment-pages[here]:
+
+. A client application  *initiates* the *GoCardless Redirect flow* with the *customer details*  (optional) and a *success page URL*.
```

Comment:
Maybe we should call out here who would be responsible for calling what:

* Webstore (e.g. ecommerce website) initiates the redirect by calling Kill Bill in the backend
* Webstore uses the redirect URL to redirect the user
* Upon redirect from GoCardless, webstore completes the flow

Basically, I'm not sure it's 100% clear who would issue the `cURL` commands from the GoCardless README.

We could also generate a diagram (a bit like https://docs.killbill.io/latest/userguide_payment.html#_hosted_payment_page_flow)? Or maybe a flowchart (https://whimsical.com/flowcharts)?